sbndaq-artdaq/Generators/ICARUS/ICARUSTriggerV3_generator.cc

```diff
@@ -0,0 +1,858 @@
+    ptime time_t_epoch(date(1970,1,1));
+    time_duration diff = t_now - time_t_epoch;
+
+    ts = diff.total_nanoseconds();
+    fNTP_time = ts;
+  }
+
+  if(fInitialStep == 0)
+  {
+    fStartOfRun = ts;
+    fInitialStep = 1;
+  }
+
+  if(data_input==""){
+    TLOG(TLVL_DEBUG) << "No data after poll with timeout? " << data_input;
+    return true;
+  }
+
+  //if shouldn't send fragments, then don't create fragment/send
+  if(generated_fragments_per_event_==0){
+    fLastEvent = fEventCounter;
+    ++fEventCounter;
+    metricMan->sendMetric("EventRate",1, "Hz", 1,artdaq::MetricMode::Rate);
+    metricMan->sendMetric("EventCounter",uint64_t{fEventCounter}, "Triggers", 1,artdaq::MetricMode::LastPoint);
+    return true;
+  }
+  
+  icarus::ICARUSTriggerInfo datastream_info = icarus::parse_ICARUSTriggerV3String(buffer);
+
+  uint64_t event_no = fEventCounter;
+  uint64_t wr_ts = datastream_info.getNanoseconds_since_UTC_epoch() + wr_time_offset_ns_;
+  if(use_wr_time_ && wr_ts > 0)
+    ts = wr_ts;
+  if(use_wr_time_ && wr_ts == 0)
+  {
+    TLOG(TLVL_WARNING) << "WR time = 0";
+  }
+  if(datastream_info.event_no > -1)
+    event_no = datastream_info.event_no;
+
+  if(datastream_info.wr_name != " WR_TS1 " || datastream_info.wr_seconds == -3 || datastream_info.wr_nanoseconds == -4)
+  {
+    TLOG(TLVL_WARNING) << "White Rabbit timestamp missing!";
+  }
+    //Add in fragment details and fragment filling function, want a fragment to contain all of the variables arriving with the trigger                                                                                    
+  //Put user variables in metadata, maybe except trigger name, try all at first and might be doing not quite correctly
+    //Only create and send fragment if the trigger number has increased, noticed can get multiple of the same trigger from the board
+  if(fLastEvent >= event_no) TLOG(TLVL_WARNING)<< "fLastEvent="<<fLastEvent << ", event_no="<<event_no;
+
+  if(fLastEvent < event_no)
+  {
+    if(fLastEvent == 0)
+    {
+      fLastTimestamp = fStartOfRun;
+      fLastTimestampBNBMaj = fStartOfRun;
+      fLastTimestampBNBMinbias = fStartOfRun;
+      fLastTimestampNuMIMaj = fStartOfRun;
+      fLastTimestampNuMIMinbias = fStartOfRun;
+      fLastTimestampOther = fStartOfRun;
+      fLastTimestampBNBOffMaj = fStartOfRun;
+      fLastTimestampBNBOffMinbias = fStartOfRun;
+      fLastTimestampNuMIOffMaj = fStartOfRun;
+      fLastTimestampNuMIOffMinbias = fStartOfRun;
+      fLastTimestampCalibMaj = fStartOfRun;
+      fLastTimestampCalibMinbias = fStartOfRun;
+    }
+
+    fDeltaGates = datastream_info.gate_id - fLastGatesNum;
+    metricMan->sendMetric("EventRate",1, "Hz", 1,artdaq::MetricMode::Rate);
+    metricMan->sendMetric("EventCounter", uint64_t(fEventCounter), "Triggers", 1,artdaq::MetricMode::LastPoint);
+    
+    if(fDeltaGates <= 0)
+      TLOG(TLVL_WARNING) << "Change in total number of beam gates for ALL <= 0!";
+
+    if(datastream_info.gate_type == 1)
```

Comment:
Do we have a documentation of the gate types and trigger types somewhere, that we could link here in the comment? I also wonder if we could replace the numbers with enums, so that the code is self-documenting?